### ISSUE TITLE
Document conflict is killing horti cli

### ISSUE THREAD

**Issue by newtewt** [medic/horticulturalist#59]
When performing an upgrade with horti running in CLI on --medic-os mode. I hit a document conflict which caused horti to crash. I have included the logs here. 
```
root@ee7d9aa81943:/horticulturalist# node src/index.js --medic-os                                                                   
Starting Horticulturalist 0.13.1 daemon in Medic OS mode                                                                              
horti:debug Running pre-boot checks +0ms                                                                                           
horti:debug Pre-boot checks OK +28ms                                                                                              
Watching for deployments                                                                                                            
Change in horti-upgrade detected                                                                                                   
 Deploying new build: medic:medic:3.8.0                                                                                                
horti:stage Horticulturalist deployment of 'medic:medic:3.8.0' initialising +0ms                                                    
horti:stage Pre-deploy cleanup +31ms                                                                                                
horti:debug Clear existing staged DBs +25s                                                                                          
horti:debug Starting compact and view cleanup +10ms                                                                                
 horti:stage Downloading and staging install +452ms                                                                                 
 horti:debug Downloading medic:medic:3.8.0, this may take some time… +442ms                                                          
horti:debug Got medic:medic:3.8.0, staging +7s                                                                                     
 horti:debug Staged as _design/:staged:medic +7s                                                                                  
   horti:stage Extracting ddocs +14s                                                                                                  
 horti:debug Storing staged: ["_design/:staged:medic-admin","_design/:staged:medic-client","_design/:staged:medic-conflicts","_design/:staged:medic-scripts","_design/:staged:medic-sms","_design/:staged:medic"] +74ms                                                  
horti:stage Warming views +9s                                                                                                      
 horti:debug Got 6 staged ddocs +9s                                                                                                
Beginning view warming                                                                                                             
 Warming views complete                                                                                                                
horti:stage View warming complete, ready to deploy +1s                                                                              
horti:stage Initiating deployment +42ms                                                                                             
horti:stage Deploying new installation +39ms                                                                                       
 horti:debug Found [{"name":"medic-api","attachmentName":"medic-api-0.1.0.tgz","digest":"md5-v/2Zh21dPN4uzUynUdyw4A=="},{"name":"medic-sentinel","attachmentName":"medic-sentinel-0.1.0.tgz","digest":"md5-BcDUBsNWuDLZNFme4rmihw=="}] +1s                              
 horti:debug Apps that are changed: [{"name":"medic-api","attachmentName":"medic-api-0.1.0.tgz","digest":"md5-v/2Zh21dPN4uzUynUdyw4A=="},{"name":"medic-sentinel","attachmentName":"medic-sentinel-0.1.0.tgz","digest":"md5-BcDUBsNWuDLZNFme4rmihw=="}] +1ms           
Unzipping changed apps to /srv/software… [ { name: 'medic-api',                                                                         
attachmentName: 'medic-api-0.1.0.tgz',                                                                                              
digest: 'md5-v/2Zh21dPN4uzUynUdyw4A==',                                                                                             
deployPath: [Function: deployPath] },                                                                                             
{ name: 'medic-sentinel',                                                                                                             
attachmentName: 'medic-sentinel-0.1.0.tgz',                                                                                         
digest: 'md5-BcDUBsNWuDLZNFme4rmihw==',                                                                                             
deployPath: [Function: deployPath] } ]                                                                                         
 Changed apps unzipped.                                                                                                              
Stopping all apps… [ 'medic-api', 'medic-sentinel' ]                                                                                 
horti:debug Stopping app: medic-api with command: sudo,-n,/boot/svc-stop,{{app}}… +4s                                             
Debug: Service 'medic-api/medic-api' exited with status 143                                                                        
 Info: Service 'medic-api/medic-api' was stopped successfully                                                                        
Success: Finished stopping services in package 'medic-api'                                                                            
horti:debug Stopped medic-api. +78ms                                                                                                
horti:debug Stopping app: medic-sentinel with command: sudo,-n,/boot/svc-stop,{{app}}… +0ms                                       
Debug: Service 'medic-sentinel/medic-sentinel' exited with status 143                                                               
Info: Service 'medic-sentinel/medic-sentinel' was stopped successfully                                                              
Success: Finished stopping services in package 'medic-sentinel'                                                                       
horti:debug Stopped medic-sentinel. +76ms                                                                                         
All apps stopped.                                                                                                                   
Deploying staged ddocs                                                                                                                
horti:debug Falling back onto getting staged ddocs in a slower, safer way +5s                                                       
horti:debug Getting list of ddocs +0ms                                                                                              
horti:debug Individually fetching _design/:staged:medic +8ms                                                                        
horti:debug Individually fetching _design/:staged:medic-admin +493ms                                                                
horti:debug Individually fetching _design/:staged:medic-client +183ms                                                               
horti:debug Individually fetching _design/:staged:medic-conflicts +18ms                                                            
 horti:debug Individually fetching _design/:staged:medic-scripts +4ms                                                                
horti:debug Individually fetching _design/:staged:medic-sms +3ms                                                                   
 horti:debug Secondary ddocs: _design/:staged:medic-admin,_design/:staged:medic-client,_design/:staged:medic-conflicts,_design/:staged:medic-scripts,_design/:staged:medic-sms +11ms                                                                                     
horti:debug Getting currently deployed secondary ddocs +0ms                                                                         
horti:debug Found 5 +7ms                                                                                                           
 horti:debug _design/medic-admin already exists, overwriting +0ms                                                                    
horti:debug _design/medic-client already exists, overwriting +0ms                                                                   
horti:debug _design/medic-conflicts already exists, overwriting +0ms                                                                
horti:debug _design/medic-scripts already exists, overwriting +0ms                                                                 
 horti:debug _design/medic-sms already exists, overwriting +0ms                                                                      
horti:debug Writing secondary ddocs +0ms                                                                                            
horti:debug Secondary ddocs written: [{"ok":true,"id":"_design/medic-admin","rev":"2-d874c21f3b8e2a4cbde7c6c0475ad441"},{"ok":true,"id":"_design/medic-client","rev":"3-38237c64974950b94b978e10fb40fb00"},{"ok":true,"id":"_design/medic-conflicts","rev":"2-3b5251fff0fd5c822de72f43a34b4342"},{"ok":true,"id":"_design/medic-scripts","rev":"2-ef490b5ba8a296880c39b89935a79ce3"},{"ok":true,"id":"_design/medic-sms","rev":"2-534b203b454a81b1e9c180b4250357ae"}] +3s                                                                       
horti:debug Primary ddoc: _design/:staged:medic +0ms                                                                               
 horti:debug Checking to see if primary exists already +0ms                                                                         
 horti:debug It does +5ms                                                                                                           
 horti:debug Writing primary ddoc +0ms                                                                                              
 horti:debug Primary ddoc written +6s                                                                                             
 Updating symlinks for changed apps… [ { name: 'medic-api',                                                                         
     attachmentName: 'medic-api-0.1.0.tgz',                                                                                        
     digest: 'md5-v/2Zh21dPN4uzUynUdyw4A==',                                                                                    
         deployPath: [Function: deployPath] },                                                                                            
 { name: 'medic-sentinel',                                                                                                            
 attachmentName: 'medic-sentinel-0.1.0.tgz',                                                                                       
  digest: 'md5-BcDUBsNWuDLZNFme4rmihw==',                                                                                        
     deployPath: [Function: deployPath] } ]                                                                                         
 Symlinks updated.                                                                                                                  
 Starting all apps… [ 'medic-api', 'medic-sentinel' ]                                                                                  
horti:debug Starting app: medic-api with command: sudo,-n,/boot/svc-start,{{app}}… +2ms                                          
 Info: Service 'medic-api/medic-api' was started successfully                                                                       
 Success: Finished starting services in package 'medic-api'                                                                            
horti:debug Started medic-api in the background. +42ms                                                                              
horti:debug Starting app: medic-sentinel with command: sudo,-n,/boot/svc-start,{{app}}… +0ms                                    
  Info: Service 'medic-sentinel/medic-sentinel' was started successfully                                                              
Success: Finished starting services in package 'medic-sentinel'                                                                      
 horti:debug Started medic-sentinel in the background. +34ms                                                                       
All apps started.                                                                                                                     
horti:stage Post-deploy cleanup, installation complete +19s                                                                         
horti:debug Deleting old medic-api from /srv/software/medic-api/md5-iz00aBf66BGsQc08Rc3k2g==… +147ms                                
horti:debug Deleting old medic-sentinel from /srv/software/medic-sentinel/md5-NT389VnR2y7QABFyLjpATQ==… +194ms                      
horti:debug Clear existing staged DBs +136ms                                                                                        
horti:debug Deleting staged ddocs: ["_design/:staged:medic","_design/:staged:medic-admin","_design/:staged:medic-client","_design/:staged:medic-conflicts","_design/:staged:medic-scripts","_design/:staged:medic-sms"] +8ms                                            
horti:debug Delete deploy ddoc +28ms                                                                                               
 horti:debug Cleanup old views +28ms                                                                                               
Watching for deployments                                                                                                            
Change in horti-upgrade detected                                                                                                   
 Deploying new build: medic:medic:3.8.1                                                                                                
horti:stage Horticulturalist deployment of 'medic:medic:3.8.1' initialising +19m                                                    
horti:stage Pre-deploy cleanup +64ms                                                                                                
horti:debug Clear existing staged DBs +19m                                                                                          
horti:debug Starting compact and view cleanup +27ms                                                                                 
horti:stage Downloading and staging install +2s                                                                                     
horti:debug Downloading medic:medic:3.8.1, this may take some time… +2s                                                            
 horti:debug Got medic:medic:3.8.1, staging +7s                                                                                      
horti:debug Staged as _design/:staged:medic +8s                                                                                    
 horti:stage Extracting ddocs +15s                                                                                                   
horti:debug Storing staged: ["_design/:staged:medic-admin","_design/:staged:medic-client","_design/:staged:medic-conflicts","_design/:staged:medic-scripts","_design/:staged:medic-sms","_design/:staged:medic"] +100ms                                                 
horti:stage Warming views +10s                                                                                                     
 horti:debug Got 6 staged ddocs +10s                                                                                               
Beginning view warming                                                                                                                
horti:debug View indexer progress +1s                                                                                               
horti:debug _design/:staged:medic-client       [4%______________________________________________________________________] +1ms      
horti:debug View indexer progress +10s                                                                                              
horti:debug _design/:staged:medic-client       [||||||||||||||||||||60%||||||||||||||||||||_____________________________] +1ms    Warming views complete                                                                                                               
 horti:debug View indexer progress +9s                                                                                              
 horti:debug _design/:staged:medic-client       [||||||||||||||||||||||||||||||||||100%||||||||||||||||||||||||||||||||||] +0ms   
 Unexpected error while warming: (Document update conflict.), continuing                                                             ********FATAL********                                                                                                               { error: 'conflict',                                                                                                                  reason: 'Document update conflict.',                                                                                                status: 409,                                                                                                                        name: 'conflict',                                                                                                                   message: 'Document update conflict.',                                                                                               docId: 'horti-upgrade' }                                  
```

**Comment by dianabarsan** [comment]
I can 100% replicate this if I edit the `horti-upgrade` doc externally manually (for example via fauxton) while view warming is progressing. The view warming code always handles conflicts successfully.
The error occurs while staging the next install step here: https://github.com/medic/horticulturalist/blob/master/src/utils.js#L123 , so it's not isolated to the view warming step. 

I think the solution is to change the way we update the doc here https://github.com/medic/horticulturalist/blob/master/src/utils.js#L134 , so it accounts for external changes and can deal with conflicts.

**Comment by garethbowen** [comment]
Added to 3.11 because it can be worked on alongside the main Angular change.

**Comment by garethbowen** [comment]
Delayed to 3.12.0 to speed up 3.11.0 release.

**Comment by mrjones-plip** [comment]
I'm getting a similar error and I want to make sure it's either this same issue or if I should file another ticket.  

I'm running mine against a local docker compose instance with zero data.  As in I run `docker-compose up` on a CHT instance that didn't exist and then immediately run `horti --install` on it.  Here's my full error.  The upgrade always completes on the CHT side, FYI:

```
COUCH_NODE_NAME=couchdb@localhost COUCH_URL=https://medic:password@192-168-68-40.my.local-ip.co/medic horti --local --install=medic:medic:3.10.0
Starting Horticulturalist 0.13.1 daemon in local mode
  horti:debug Running pre-boot checks +0ms
  horti:debug Pre-boot checks OK +124ms
Doing install to medic:medic:3.10.0
  horti:debug Bootstrapping upgrade doc +21ms
  horti:debug Starting app: medic-api with command: horti-svc-start,/home/mrjones/.horticulturalist/deployments,{{app}}… +39ms
[/usr/bin/horti-svc-start] Starting medic-api from /home/mrjones/.horticulturalist/deployments/medic-api/current...
[/usr/bin/horti-svc-start] medic-api started.
/usr/bin/horti-svc-start: line 8: cd: /home/mrjones/.horticulturalist/deployments/medic-api/current: No such file or directory
  horti:debug Started medic-api in the background. +10ms
  horti:debug Starting app: medic-sentinel with command: horti-svc-start,/home/mrjones/.horticulturalist/deployments,{{app}}… +0ms
[/usr/bin/horti-svc-start] Starting medic-sentinel from /home/mrjones/.horticulturalist/deployments/medic-sentinel/current...
[/usr/bin/horti-svc-start] medic-sentinel started.
/usr/bin/horti-svc-start: line 8: cd: /home/mrjones/.horticulturalist/deployments/medic-sentinel/current: No such file or directory
  horti:debug Started medic-sentinel in the background. +7ms
Deploying new build: medic:medic:3.10.0
********FATAL********
{
  error: 'conflict',
  reason: 'Document update conflict.',
  status: 409,
  name: 'conflict',
  message: 'Document update conflict.',
  docId: 'horti-upgrade'
}
+ app=medic-api
+ echo '[/usr/bin/horti-svc-stop] Stopping medic-api...'
[/usr/bin/horti-svc-stop] Stopping medic-api...
+ pgrep -f HORTICULTURALIST_APP=medic-api
+ xargs kill
+ app=medic-sentinel
+ echo '[/usr/bin/horti-svc-stop] Stopping medic-sentinel...'
[/usr/bin/horti-svc-stop] Stopping medic-sentinel...
+ pgrep -f HORTICULTURALIST_APP=medic-sentinel
+ xargs kill
                                                                                                                                                                                                   
Usage:
 kill [options] <pid> [...]

Options:
 <pid> [...]            send signal to every <pid> listed
 -<signal>, -s, --signal <signal>
                        specify the <signal> to be sent
 -l, --list=[<signal>]  list all signal names, or convert one to a name
 -L, --table            list all signal names in a nice table

 -h, --help     display this help and exit
 -V, --version  output version information and exit

For more details see kill(1).
+ true
+ echo '[/usr/bin/horti-svc-stop] medic-api stopped.'
[/usr/bin/horti-svc-stop] medic-api stopped.

Usage:
 kill [options] <pid> [...]

Options:
 <pid> [...]            send signal to every <pid> listed
 -<signal>, -s, --signal <signal>
                        specify the <signal> to be sent
 -l, --list=[<signal>]  list all signal names, or convert one to a name
 -L, --table            list all signal names in a nice table

 -h, --help     display this help and exit
 -V, --version  output version information and exit

For more details see kill(1).
+ true
+ echo '[/usr/bin/horti-svc-stop] medic-sentinel stopped.'
[/usr/bin/horti-svc-stop] medic-sentinel stopped.
```

**Comment by dianabarsan** [comment]
I think you're getting the same error but at a different stage: after the upgrade was successfully completed and deployed. It can happen at any stage, however, and interrupt the process.

**Comment by mrjones-plip** [comment]
Ah - good good - as I suspected! Thanks for confirming @dianabarsan .

**Comment by latin-panda** [comment]
I got this issue while following the steps to[ install FR](https://docs.communityhealthtoolkit.org/core/releases/feature_releases/#initial-fr-installation).
![Screen Shot 2022-08-26 at 9 49 22 am](https://user-images.githubusercontent.com/66472237/187174826-6c401629-55c6-4cc8-9f6d-73b866c2075f.png)

**Comment by mrjones-plip** [comment]
@latin-panda  - I was just working with @tatilepizs  who was getting the same error just now.  We were able to "fix" it because I remembered [this comment](https://github.com/medic/horticulturalist/issues/59#issuecomment-885289478) from above:

> I'm running mine against a local docker compose instance with zero data. As in I run docker-compose up on a CHT instance that didn't exist and then immediately run horti --install on it. Here's my full error. **The upgrade always completes on the CHT side**

tl;dr - you get the error - but check the admin GUI, the upgrade should start and then you should be prompted to "reload application" when it completes

**Comment by dianabarsan** [comment]
Removing Help wanted as Horticulturalist is in maintenance mode.
